File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueIdaEVolta.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import java.util.ArrayList;
import java.util.List;

/**
 * Não é bem um "ataque" - é a checagem de sanidade básica: encrypt seguido de
 * decrypt precisa devolver o texto original, em qualquer tamanho.
 */
public class AtaqueIdaEVolta implements AtaqueInterface {
    private final int tamanhoMaximo;

    public AtaqueIdaEVolta() {
        this(130);
    }

    public AtaqueIdaEVolta(int tamanhoMaximo) {
        this.tamanhoMaximo = tamanhoMaximo;
    }

    @Override
    public String nome() {
        return "Ida-e-volta (round trip)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        List<String> falhas = new ArrayList<>();
        for (int len = 0; len <= this.tamanhoMaximo; len++) {
            String texto = "Q".repeat(len);
            try {
                String token = alvo.encrypt(texto);
                String decifrado = alvo.decrypt(token);
                if (!decifrado.equals(texto)) {
                    falhas.add(String.valueOf(len));
                }
            } catch (Exception e) {
                String mensagem = e.getMessage() != null ? e.getMessage() : e.toString();
                falhas.add(len + " (erro: " + mensagem + ")");
            }
        }

        if (!falhas.isEmpty()) {
            return new ResultadoAtaque(
                    nome(),
                    true,
                    Severidade.CRITICA,
                    "Falhou em " + falhas.size() + " tamanho(s): "
                            + String.join(", ", falhas.subList(0, Math.min(10, falhas.size()))));
        }

        return new ResultadoAtaque(
                nome(),
                false,
                Severidade.INFO,
                "Todos os " + (this.tamanhoMaximo + 1) + " tamanhos (0 a " + this.tamanhoMaximo + ") OK");
    }
}

File: packages/go/internal/seguranca/ataques/ataque_ida_e_volta.go
package ataques

import (
	"fmt"
	"strings"

	"criptografia/internal/seguranca"
)

// AtaqueIdaEVolta é a checagem de sanidade básica: encrypt seguido de decrypt
// precisa devolver o texto original, em qualquer tamanho.
type AtaqueIdaEVolta struct {
	tamanhoMaximo int
}

// NovoAtaqueIdaEVolta cria o ataque. Sem argumento, usa tamanhoMaximo = 130.
func NovoAtaqueIdaEVolta(tamanhoMaximo ...int) *AtaqueIdaEVolta {
	t := 130
	if len(tamanhoMaximo) > 0 {
		t = tamanhoMaximo[0]
	}
	return &AtaqueIdaEVolta{tamanhoMaximo: t}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueIdaEVolta) Nome() string {
	return "Ida-e-volta (round trip)"
}

// Executar roda o ataque.
func (a *AtaqueIdaEVolta) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	falhas := []string{}

	for length := 0; length <= a.tamanhoMaximo; length++ {
		texto := strings.Repeat("Q", length)
		token, err := alvo.Encrypt([]byte(texto))
		if err != nil {
			falhas = append(falhas, fmt.Sprintf("%d (erro: %s)", length, err.Error()))
			continue
		}
		decifrado, err := alvo.Decrypt(token)
		if err != nil {
			falhas = append(falhas, fmt.Sprintf("%d (erro: %s)", length, err.Error()))
			continue
		}
		if decifrado != texto {
			falhas = append(falhas, fmt.Sprintf("%d", length))
		}
	}

	if len(falhas) > 0 {
		limite := min(10, len(falhas))
		return seguranca.ResultadoAtaque{
			NomeAtaque: a.Nome(),
			Vulneravel: true,
			Severidade: seguranca.SevCritica,
			Detalhes:   "Falhou em " + fmt.Sprint(len(falhas)) + " tamanho(s): " + strings.Join(falhas[:limite], ", "),
			Dados:      map[string]any{"falhas": falhas},
		}, nil
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: false,
		Severidade: seguranca.SevInfo,
		Detalhes:   fmt.Sprintf("Todos os %d tamanhos (0 a %d) OK", a.tamanhoMaximo+1, a.tamanhoMaximo),
	}, nil
}
